src/GraphQL/Server.hs

```diff
@@ -180,27 +196,37 @@ instance forall v. ReadValue v => ReadValue (Maybe v) where
 -- the name matches the query. Note that the name is *not* in monad m,
 -- but the value is. This is necessary so we can skip execution if the
 -- name doesn't match.
-data NamedFieldExecutor m = NamedFieldExecutor AST.Name (m GValue.Value)
-
-executeNamedField :: Monad m => AST.Field -> NamedFieldExecutor m -> m (Maybe GValue.ObjectField)
-executeNamedField (AST.Field alias name _ _ _) (NamedFieldExecutor k mValue)
-  | name == k = do
-      value <- mValue
-      let name' = fromMaybe name alias
-      pure (Just (GValue.ObjectField name' value))
-  | otherwise = pure Nothing
-
--- Deconstruct object type signature and handler value at the same
--- time and run type-directed code for each field.
-resolveField :: forall a (m :: Type -> Type). BuildFieldResolver m a => m GValue.ObjectField -> FieldHandler m a -> AST.Selection -> m GValue.ObjectField
-resolveField fallback lh selection@(AST.SelectionField field) = do
-  let fieldExecutor = buildFieldResolver @m @a lh selection
-  case fieldExecutor of
-    Left err -> throwM err
-    Right executor -> do
-      objectField <- executeNamedField field executor
-      maybe fallback pure objectField
-resolveField _ _ f = queryError ("Unexpected Selection value. Is the query normalized?: " <> show f)
+data NamedValueResolver m = NamedValueResolver AST.Name (m Result)
+
+
+
+-- Iterate through handlers (zipped together with their type
+-- definition) and execute handler if the name matches.
+--
+-- TODO: tuple return not great (but either is wrong, too because it discards errors.
+type ResolveFieldResult = ([QueryError], Maybe GValue.ObjectField)
+resolveField :: forall a (m :: Type -> Type). BuildFieldResolver m a
+  => FieldHandler m a -> m ResolveFieldResult -> AST.Selection -> m ResolveFieldResult
+resolveField handler nextHandler selection@(AST.SelectionField (AST.Field _ queryFieldName _ _ _)) =
+  case buildFieldResolver @m @a handler selection of
+    -- TODO the fact that this doesn't fit together nicely makes me think that ObjectField is not a good idea)
+    Left err -> pure ([err], Just (GValue.ObjectField queryFieldName GValue.ValueNull))
+    Right (NamedValueResolver name' resolver) -> runResolver name' resolver
+  where
+    runResolver :: AST.Name -> m Result -> m ResolveFieldResult
+    runResolver name' resolver
+      | queryFieldName == name' = do
+          Result errs value <- resolver
+          pure (errs, Just (GValue.ObjectField queryFieldName value))
+      | otherwise = nextHandler
+resolveField _ _ f =
+  -- TODO, I'm not sure whether this should be a panic instead? (If we
+  -- panic we can avoid the Maybe in the ResolveFieldResult type).
+  --
+  -- An alternative intepretation is that we can formulate
+  -- syntactically valid queries that are still semantically broken in
+  -- which case we should not panic.
+  pure ([InternalError ("Unexpected Selection value. Is the query normalized?: " <> show f)], Nothing)
```

Comment:
This can maybe wait for a second PR, but I'd like to see us generating more structured errors in this code (e.g. `SchemaError (UnexpectedSelection f)`) and then using something like `formatNameError` to generate user-friendly messages. Probably we'd have a typeclass with a `formatError` method, or something.

The reason is that it helps to reason about all the possible things that can go wrong.

If we made this change, we'd see that `QueryError` (or `Error`, if we rename it such), would have very very few `Text` fields.